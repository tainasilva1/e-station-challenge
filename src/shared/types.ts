export interface IChartData {
  [key: string]: number | string;
}

export type DataKey = {
  key: string;
  color: string;
}

export interface IConsumptionData {
  id: string,
  reference: string,
  year: number,
  month: number,
  day: number,
  hour: number,
  consumption: number,
  agent: string,
  meter: string,
  origin: string,
  intervalInMinutes: number,
  isEstimated: boolean
}
