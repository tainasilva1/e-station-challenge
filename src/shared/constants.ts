export const MONTH_NAMES = ['Jan', 'Fev', 'Mar', 'Abr', 'Mai', 'Jun', 'Jul', 'Ago', 'Set', 'Out', 'Nov', 'Dez']

export const KEYS_DATA_BY_DAY = {
  CONSUMPTION: 'consumption',
  MAXIMUM_FLEX: 'maximum_flex',
  MINIMUM_FLEX: 'minimum_flex',
  FLAT_CONSUMPTION: 'flat_consumption'
}

export const KEYS_BY_LAST_WEEK = [{ key: 'consumo', color: '#8884d8'}]

export const KEYS_BY_YEAR = [
  { key: '2021', color: '#3c81f6' },
  { key: '2022', color: '#1ab8a6' }
]

export const KEYS_BY_DATE = [
  {
    key: KEYS_DATA_BY_DAY.CONSUMPTION,
    color: '#5e9ae5'
  },
  {
    key: KEYS_DATA_BY_DAY.MAXIMUM_FLEX,
    color: '#f97415'
  },
  {
    key: KEYS_DATA_BY_DAY.MINIMUM_FLEX,
    color: '#6466f1'
  },
  {
    key: KEYS_DATA_BY_DAY.FLAT_CONSUMPTION,
    color: '#23c45e'
  },
]


export const BY_DAY_LEGENDS = {
  [KEYS_DATA_BY_DAY.CONSUMPTION]: 'Consumo',
  [KEYS_DATA_BY_DAY.MAXIMUM_FLEX]: 'Flex Máxima',
  [KEYS_DATA_BY_DAY.MINIMUM_FLEX]: 'Flex Mínima',
  [KEYS_DATA_BY_DAY.FLAT_CONSUMPTION]: 'Consumo Flat'
}

export const CONSUMPTION_KEYS = {
  AGENT: 'agent',
  METER: 'meter',
  REFERENCE: 'reference',
  HOUR: 'hour',
  CONSUMPTION: 'consumption',
  ORIGIN: 'origin'
}

export const CONSUMPTION_LABELS = {
  [CONSUMPTION_KEYS.AGENT]: 'Agente',
  [CONSUMPTION_KEYS.METER]: 'Ponto',
  [CONSUMPTION_KEYS.REFERENCE]: 'Data',
  [CONSUMPTION_KEYS.HOUR]: 'Hora',
  [CONSUMPTION_KEYS.CONSUMPTION]: 'Consumo Ativo (MWh)',
  [CONSUMPTION_KEYS.ORIGIN]: 'Origem'
}