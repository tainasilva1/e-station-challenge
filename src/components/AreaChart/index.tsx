import { DataKey, IChartData } from '@/src/shared/types';
import { memo } from 'react';
import { AreaChart as AreaChartRecharts, XAxis, YAxis, CartesianGrid, Tooltip, Legend, Line, Area, ResponsiveContainer } from 'recharts';
import ContentLoader from 'react-content-loader';

type Props = {
  data?: IChartData[];
  dataKey: DataKey[];
  unit?: string;
}

export const AreaChart = ({
  data,
  dataKey,
  unit = '(MWh)',
}: Props) => {
  if (!data) return (
    <ContentLoader
      width='100%'
      height={350}
    >
      <path d="M 75 229.375 h 6.377 c 2.126 0 4.251 -17.125 6.377 -25.687 C 89.88 195.125 92.005 178 94.131 178 c 2.126 0 4.251 94.188 6.377 94.188 c 2.126 0 4.251 -119.875 6.377 -119.875 c 2.125 0 4.251 85.625 6.377 85.625 c 2.125 0 4.251 -34.25 6.377 -34.25 c 2.125 0 4.251 85.625 6.377 85.625 c 2.125 0 4.251 -171.25 6.377 -171.25 c 2.125 0 4.251 119.875 6.377 119.875 c 2.125 0 4.251 -34.25 6.377 -51.375 c 2.125 -17.125 4.251 -51.375 6.377 -51.375 c 2.125 0 4.251 128.437 6.377 128.437 c 2.125 0 4.251 -97.042 6.376 -102.75 c 2.126 -5.708 4.252 -2.854 6.377 -8.562 c 2.126 -5.709 4.252 -59.938 6.377 -59.938 c 2.126 0 4.252 102.75 6.377 102.75 c 2.126 0 4.252 -51.375 6.377 -51.375 c 2.126 0 4.252 24.26 6.377 34.25 c 2.126 9.99 4.252 25.688 6.377 25.688 c 2.126 0 4.252 -34.25 6.377 -34.25 c 2.126 0 4.252 31.395 6.377 51.375 c 2.126 19.979 4.252 68.5 6.377 68.5 c 2.126 0 4.251 -17.125 6.377 -42.813 c 2.126 -25.687 4.251 -111.312 6.377 -111.312 c 2.126 0 4.251 25.687 6.377 25.687 c 2.126 0 4.251 -25.687 6.377 -25.687 c 2.126 0 4.251 111.312 6.377 111.312 c 2.126 0 4.251 -42.812 6.377 -42.812 c 2.126 0 4.251 94.187 6.377 94.187 c 2.126 0 4.251 -102.75 6.377 -102.75 c 2.126 0 4.251 27.115 6.377 42.813 c 2.125 15.697 4.251 51.375 6.377 51.375 c 2.125 0 4.251 -17.125 6.377 -34.25 c 2.125 -17.125 4.251 -68.5 6.377 -68.5 c 2.125 0 4.251 94.187 6.377 94.187 h 6.377 c 2.125 0 4.251 -188.375 6.377 -188.375 h 6.377 c 2.125 0 4.251 37.104 6.377 59.938 c 2.125 22.833 4.251 77.062 6.376 77.062 c 2.126 0 4.252 -145.562 6.377 -145.562 c 2.126 0 4.252 29.968 6.377 59.937 c 2.126 29.969 4.252 119.875 6.377 119.875 h 6.377 c 2.126 0 4.252 -179.812 6.377 -179.812 c 2.126 0 4.252 77.062 6.377 77.062 c 2.126 0 4.252 -42.812 6.377 -42.812 c 2.126 0 4.252 18.552 6.377 42.812 c 2.126 24.26 4.252 102.75 6.377 102.75 c 2.126 0 4.251 -68.5 6.377 -94.187 c 2.126 -25.688 4.251 -59.938 6.377 -59.938 c 2.126 0 4.251 39.958 6.377 59.938 c 2.126 19.979 4.251 59.937 6.377 59.937 c 2.126 0 4.251 -94.187 6.377 -94.187 c 2.126 0 4.251 17.125 6.377 42.812 c 2.126 25.688 4.251 111.313 6.377 111.313 c 2.126 0 4.251 -145.563 6.377 -145.563 c 2.126 0 4.251 17.125 6.377 17.125 c 2.125 0 4.251 -51.375 6.377 -51.375 c 2.125 0 4.251 8.563 6.377 25.688 c 2.125 17.124 4.251 145.562 6.377 145.562 c 2.125 0 4.251 -137 6.377 -137 c 2.125 0 4.251 68.5 6.377 68.5 c 2.125 0 4.251 -94.187 6.377 -94.187 h 6.377 c 2.125 0 4.251 11.416 6.377 34.25 c 2.125 22.833 4.251 111.312 6.377 111.312 c 2.125 0 4.251 -59.937 6.376 -59.937 c 2.126 0 4.252 17.125 6.377 17.125 c 2.126 0 4.252 -97.042 6.377 -102.75 c 2.126 -5.709 4.252 -8.563 6.377 -8.563 c 2.126 0 4.252 137 6.377 137 c 2.126 0 4.252 -17.125 6.377 -17.125 c 2.126 0 4.252 17.125 6.377 17.125 c 2.126 0 4.252 -162.687 6.377 -162.687 c 2.126 0 4.252 137 6.377 137 c 2.126 0 4.251 -42.813 6.377 -42.813 c 2.126 0 4.251 77.063 6.377 77.063 c 2.126 0 4.251 -68.5 6.377 -68.5 c 2.126 0 4.251 15.697 6.377 34.25 c 2.126 18.552 4.251 77.062 6.377 77.062 c 2.126 0 4.251 -119.875 6.377 -119.875 c 2.126 0 4.251 85.625 6.377 85.625 c 2.126 0 4.251 -179.812 6.377 -179.812 c 2.126 0 4.251 38.531 6.377 51.375 c 2.126 12.843 4.251 8.562 6.377 25.687 c 2.125 17.125 4.251 77.063 6.377 77.063 c 2.125 0 4.251 -119.875 6.377 -119.875 c 2.125 0 4.251 119.875 6.377 119.875 c 2.125 0 4.251 -17.125 6.377 -17.125 c 2.125 0 4.251 34.25 6.377 34.25 c 2.125 0 4.251 -88.48 6.377 -111.313 c 2.125 -22.833 4.251 -34.25 6.377 -34.25 c 2.125 0 4.251 51.375 6.377 51.375 c 2.125 0 4.251 -25.687 6.376 -25.687 c 2.126 0 4.252 68.5 6.377 68.5 c 2.126 0 4.252 -51.375 6.377 -51.375 c 2.126 0 4.252 108.458 6.377 119.875 c 2.126 11.416 4.252 17.125 6.377 17.125 c 2.126 0 4.252 -119.875 6.377 -119.875 c 2.126 0 4.252 85.625 6.377 85.625 c 2.126 0 4.252 -42.813 6.377 -42.813 c 2.126 0 4.252 77.063 6.377 77.063 c 2.126 0 4.251 -7.136 6.377 -17.125 c 2.126 -9.99 4.251 -42.813 6.377 -42.813 c 2.126 0 4.251 25.688 6.377 25.688 c 2.126 0 4.251 -5.709 6.377 -17.125 c 2.126 -11.417 4.251 -77.063 6.377 -77.063 c 2.126 0 4.251 2.854 6.377 8.563 c 2.126 5.708 4.251 77.062 6.377 77.062 c 2.126 0 4.251 -111.312 6.377 -111.312 c 2.126 0 4.251 77.062 6.377 77.062 c 2.126 0 4.251 -51.375 6.377 -51.375 c 2.125 0 4.251 2.854 6.377 8.563 c 2.125 5.708 4.251 85.625 6.377 85.625 h 6.377 c 2.125 0 4.251 -102.75 6.377 -119.875 c 2.125 -17.125 4.251 -25.688 6.377 -25.688 c 2.125 0 4.251 77.063 6.377 77.063 c 2.125 0 4.251 -34.25 6.377 -34.25 c 2.125 0 4.251 2.854 6.377 8.562 c 2.125 5.708 4.251 85.625 6.376 85.625 c 2.126 0 4.252 -59.937 6.377 -59.937 c 2.126 0 4.252 42.812 6.377 42.812 c 2.126 0 4.252 -111.312 6.377 -111.312 c 2.126 0 4.252 2.854 6.377 8.562 c 2.126 5.708 4.252 145.563 6.377 145.563 c 2.126 0 4.252 -57.084 6.377 -77.063 c 2.126 -19.979 4.252 -42.812 6.377 -42.812 c 2.126 0 4.252 119.875 6.377 119.875 c 2.126 0 4.252 -45.667 6.377 -68.5 c 2.126 -22.834 4.251 -68.5 6.377 -68.5 c 2.126 0 4.251 71.354 6.377 77.062 c 2.126 5.708 4.251 8.563 6.377 8.563 c 2.126 0 4.251 -94.188 6.377 -94.188 c 2.126 0 4.251 59.938 6.377 59.938 c 2.126 0 4.251 -85.625 6.377 -85.625 c 2.126 0 4.251 179.812 6.377 179.812 c 2.126 0 4.251 -111.312 6.377 -111.312 c 2.126 0 4.251 77.062 6.377 77.062 c 2.125 0 4.251 -128.437 6.377 -128.437 c 2.125 0 4.251 77.062 6.377 77.062 c 2.125 0 4.251 -42.812 6.377 -42.812 c 2.125 0 4.251 85.625 6.377 85.625 c 2.125 0 4.251 -68.5 6.377 -68.5 c 2.125 0 4.251 34.25 6.377 34.25 c 2.125 0 4.251 -102.75 6.377 -102.75 c 2.125 0 4.251 85.624 6.377 102.75 c 2.125 17.125 4.251 9.989 6.377 25.687 c 2.125 15.698 4.251 68.5 6.376 68.5 c 2.126 0 4.252 -85.625 6.377 -85.625 c 2.126 0 4.252 25.688 6.377 25.688 c 2.126 0 4.252 -5.709 6.377 -17.125 c 2.126 -11.417 4.252 -94.188 6.377 -94.188 c 2.126 0 4.252 15.698 6.377 25.688 c 2.126 9.989 4.252 34.25 6.377 34.25 c 2.126 0 4.252 -59.938 6.377 -59.938 c 2.126 0 4.252 52.802 6.377 77.063 c 2.126 24.26 4.251 68.5 6.377 68.5 c 2.126 0 4.251 -15.698 6.377 -25.688 c 2.126 -9.99 4.251 -14.271 6.377 -34.25 c 2.126 -19.979 4.251 -85.625 6.377 -85.625 c 2.126 0 4.251 128.438 6.377 128.438 c 2.126 0 4.251 -85.625 6.377 -102.75 c 2.126 -17.125 4.251 -25.688 6.377 -25.688 c 2.126 0 4.251 68.5 6.377 68.5 c 2.126 0 4.251 -51.375 6.377 -51.375 h 6.377 c 2.125 0 4.251 111.313 6.377 111.313 c 2.125 0 4.251 -162.688 6.377 -162.688 c 2.125 0 4.251 37.104 6.377 68.5 c 2.125 31.396 4.251 119.875 6.377 119.875 c 2.125 0 4.251 -77.062 6.377 -77.062 c 2.125 0 4.251 68.5 6.377 68.5 c 2.125 0 4.251 -51.375 6.377 -51.375 c 2.125 0 4.251 19.979 6.377 25.687 c 2.125 5.708 4.251 8.563 6.376 8.563 c 2.126 0 4.252 -77.063 6.377 -77.063 c 2.126 0 4.252 85.625 6.377 85.625 c 2.126 0 4.252 -105.604 6.377 -128.437 c 2.126 -22.834 4.252 -34.25 6.377 -34.25 c 2.126 0 4.252 59.937 6.377 59.937 c 2.126 0 4.252 -34.25 6.377 -34.25 h 6.377 c 2.126 0 4.252 11.417 6.377 34.25 c 2.126 22.833 4.251 137 6.377 137 c 2.126 0 4.251 -179.812 6.377 -179.812 c 2.126 0 4.251 119.875 6.377 119.875 c 2.126 0 4.251 -17.125 6.377 -17.125 c 2.126 0 4.251 8.562 6.377 17.125 V 315 H 75 z" />
    </ContentLoader>
  )
  return (
    <ResponsiveContainer width="100%" height={350}>
      <AreaChartRecharts data={data}>
        <CartesianGrid strokeDasharray="3 3" vertical={false} />
        <XAxis style={{ fontSize: '0.8rem' }} tickLine={false} dataKey="name" minTickGap={8} />
        <YAxis width={70} style={{ fontSize: '0.8rem', wordBreak: 'break-all', maxWidth: '60px' }} unit={unit} tickLine={false} domain={[88, 120]} />
        <Tooltip formatter={(value, name) => [`${value} ${unit}`, name]} />
        <Legend align='right' verticalAlign='top' iconType='circle' height={36} />
        {dataKey.map((data) => (
          <Area key={data.key} type="monotone" dataKey={data.key} stroke={data.color} strokeWidth={2} fill={data.color} fillOpacity={0.1} />
        ))}
      </AreaChartRecharts>
    </ResponsiveContainer>
  )
}
export default memo(AreaChart);