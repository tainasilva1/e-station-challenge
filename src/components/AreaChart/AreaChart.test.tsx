import { KEYS_BY_LAST_WEEK } from '@/src/shared/constants';
import { render } from '@testing-library/react'
import AreaChart from '../AreaChart';

const dataMock = [{ "name": "24/12", "consumo": 98 }, { "name": "24/12", "consumo": 98 }, { "name": "24/12", "consumo": 101 }, { "name": "24/12", "consumo": 104 }, { "name": "24/12", "consumo": 93 }, { "name": "24/12", "consumo": 107 }, { "name": "24/12", "consumo": 97 }, { "name": "24/12", "consumo": 101 }, { "name": "24/12", "consumo": 91 }, { "name": "24/12", "consumo": 111 }, { "name": "24/12", "consumo": 97 }]
test('renders AreaChart correctly', () => {
  const { container } = render(<AreaChart data={dataMock} dataKey={KEYS_BY_LAST_WEEK} />)
  expect(container).toMatchSnapshot()
})