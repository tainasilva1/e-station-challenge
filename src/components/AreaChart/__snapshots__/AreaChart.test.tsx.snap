// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders AreaChart correctly 1`] = `
<div>
  <div
    class="recharts-responsive-container"
    style="width: 100%; height: 350px; min-width: 0;"
  />
</div>
`;
